src/execution/window_segment_tree.cpp

```diff
@@ -534,11 +507,17 @@ void WindowSegmentTree::ConstructTree() {
 }
 
 void WindowSegmentTree::Evaluate(WindowAggregatorState &lstate, const idx_t *begins, const idx_t *ends, Vector &result,
-                                 idx_t count) const {
+                                 idx_t count, idx_t row_idx) const {
```

Comment:
I suspect what is needed here is for most of the body of `Evaluate` to become an inner utility (`EvaluateInternal`?) that gets passed a state vector for accumulation? Then after it evaluates each part, it combines them into the output states and finalizes. In the (common case) where there are no exclusions, it just calls `EvaluateInternal` directly with the output states. After both parts have been evaluated, it takes a final pass to set `NULL` results.

`EvaluateInternal` will also need to know what `FramePart` it is dealing with, but if you do it this way, the `enum` just becomes local to this class, instead of global.